clang/tools/3c/README.md

```diff
@@ -0,0 +1,65 @@
+
+As an additional safeguard, `f.checked.c` or `f.checked.h` is not written if it is outside the _base directory_, which defaults to the working directory but can be overridden with the `-base-dir` flag.  This can help ensure that you don't unintentionally modify external libraries, though of course if their header files are not annotated for Checked C, your ability to convert your own program to Checked C may be limited.  (See [below](#annotated-system-headers) about system headers.)  However, there is a bug in the way the base directory check handles `..` path components in `-I` directories and `#include` paths (terrible, we know!), so until we can fix the bug, please avoid using `..` path components (e.g., use `-I` with an absolute path instead).
+
+You can ignore the errors about a compilation database not being found.  You can specify a single set of flags to use for all files by prefixing them with `-extra-arg-before=`, for example:
+
+```
+3c -alltypes -output-postfix=checked -extra-arg-before=-Isome/include/path foo.c bar.c
+```
+
+(If you were using a compilation database, such "extra" flags would be added to any flags in the database.)
+
+The `-alltypes` option causes `3c` to try to infer array types.  We want to make this the default but haven't done so yet because it breaks some things.
+
+## Annotated system headers
+
+As mentioned above, if a Checked C program uses an external library, the library's headers must have Checked C annotations on the declarations of the elements (functions, variables, types, etc.) used by the program in order for the program to access those items safely.  Elements with unannotated declarations can only be accessed unsafely.
+
+In the future, we may have some scheme analogous to [DefinitelyTyped](https://definitelytyped.org/) to distribute Checked C headers for existing C libraries.  In the meantime, the Checked C project maintains annotated versions of the most common "system" header files (`stdio.h`, etc.).  The annotated header files do not include all elements, but they `#include` your system's original header files, so your program can still use unannotated elements unsafely.  If you followed the [build instructions](../../docs/checkedc/3C/README.md), these header files should be present in `llvm/projects/checkedc-wrapper/checkedc/include`.
+
+Currently, the annotated header files are named with a `_checked.h` suffix, e.g., `stdio_checked.h`, so that `stdio_checked.h` can `#include <stdio.h>` without causing infinite recursion.  We hope to switch to `#include_next` and remove the suffix soon.  In the meantime, you have to modify your code to `#include <stdio_checked.h>` instead of `stdio.h` and so forth.  The `clang/tools/3c/utils/update-includes.py` tool will do this for you.  It takes one argument: the name of a file containing a list of paths of `.c` and `.h` files to be updated.  In the example of the previous section, you would run:
```

Comment:
All of this is standard Checked C; is there something in their repository we can reference about Checked headers?

Reply:
Agreed, but I could only find [this explanation of what the headers are](https://github.com/Microsoft/checkedc-clang/wiki/Checked-C-clang-user-manual#header-files) and [this old instruction to run update-includes.py that doesn't give any details](https://github.com/Microsoft/checkedc/wiki/Legacy-Conversion-Tips#convert-included-headers-first).  Both of these are in the wiki, and I don't have permission to edit it; there's nothing in the repo.

Shall we keep my text for the Microsoft PR?

I noticed we should probably update [the reference to 3C](https://github.com/Microsoft/checkedc/wiki/Legacy-Conversion-Tips#optionally-make-use-of-automated-conversion-tools) ASAP, but not for our current deadline.

Reply:
OK to keep your text, but can you also include a link the Microsoft wiki from it, no?

Reply:
Added the first wiki link.  I don't think the second has any value.  OK?

Reply:
Yes.